src/Parsers/ASTCollation.cpp

```diff
@@ -0,0 +1,21 @@
+#include <Parsers/ASTCollation.h>
+
+namespace DB
+{
+    ASTPtr ASTCollation::clone() const
+    {
+        auto res = std::make_shared<ASTCollation>(*this);
+        res->collation = collation;
```

Comment:
Shouldn't we clone recursively?
```suggestion
        res->collation = collation->clone();
```